README.md
# Courtney UI (Botfront)

![Image](https://img.shields.io/npm/v/rasa-webchat.svg 'icon')

## Rasa Webchat 💬

---

You can download the source code from [NPM](https://www.npmjs.com/package/botfront) or [GitHub](https://github.com/botfront/rasa-webchat)

Version 1.0.1 of the rasa webchat is made to work with version 2.3.x and 2.4.x of rasa. Use 1.0.0 for other rasa versions.

A chat widget to deploy virtual assistants made with [Rasa](https://github.com/rasaHQ/rasa) or [Botfront](https://github.com/botfront/botfront?utm_source=rasa_webchat) on any website.

![Image](./assets/rasa_webchat.gif 'icon')

---

## Features

-   Text Messages
-   Quick Replies
-   Images
-   Carousels
-   Markdown support
-   Persistent sessions
-   Typing indications
-   Smart delay between messages
-   Easy to import in a script tag or as a React Component

---

## Prerequisite

-   Node version 12.16.1

---

## Steps To Run

#### Continuous Integration/Install Command:

```bash
$ npm ci
```

#### DEV Deployment Steps:

```bash
$ npm run dev
```

#### PROD Deployment Steps:

```bash
$ npm run build

$ cd lib

$ python3 -m http.server PORT_NUMBER
```

---

## Using Docker

Docker is an open platform for developing, shipping, and running applications. Docker enables you to separate your applications from your infrastructure so you can deliver software quickly.

## One step method:

```
$ cd deployment
$ bash dev_deployment.sh
```

---

## Step by step method:

### To build the images:

For development:

```
$ docker-compose -f docker-compose-local.yml build
```

For production:

```
$ docker-compose build
```

### To spin up containers from the images use:

For development:

```
$ docker-compose -f docker-compose-local.yml up
```

### To push the images to the repository:

Before pushing make sure to login to the account that contains the repository. To login use:

```
$ docker login <account_name>
```

The user will be prompted for the username and password.

For development:

```
$ docker-compose -f docker-compose-local.yml push
```

For production:

```
$ docker-compose push
```

---

## Usage

### In a `<script>` tag

In your `<body/>`:

```html
<script>
    !(function () {
        let e = document.createElement('script'),
            t = document.head || document.getElementsByTagName('head')[0];
        (e.src = 'https://cdn.jsdelivr.net/npm/rasa-webchat@1.0.1/lib/index.js'), // Replace BOTFRONT URL
            (e.async = !0),
            (e.onload = () => {
                window.WebChat.default(
                    {
                        initPayload: '/welcome', // welcome payload for proactive message
                        customData: { language: 'en' }, // please refer customData below
                        socketUrl: 'http://localhost:5005/', // RASA URL
                    },
                    null
                );
            }),
            t.insertBefore(e, t.firstChild);
    })();
</script>
```

#### customData Reference:

```html
<script>
    {
    	customData: {
    		language: "en",
    		user: {
                "id": "",
                "name": ""
    		}, // Sender Details
            authUrl: '' // COURTney REST API Endpoint
    	}
    }
</script>
```

---

### As a React component

Install the [npm package](https://npmjs.com/rasa-webchat):

```bash
$ npm install rasa-webchat
```

Then:

```javascript
import Widget from 'rasa-webchat';

function CustomWidget = () => {
  return (
    <Widget
      initPayload={"/get_started"}
      socketUrl={"http://localhost:5500"}
      socketPath={"/socket.io/"}
      customData={{"language": "en"}} // arbitrary custom data. Stay minimal as this will be added to the socket
      title={"Title"}
    />
  )
}
```

-   Make sure to have the prop `embedded`
    set to `true` if you don't want to see the launcher.

---

## Parameters

| Prop / Param           | Default value      | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
| ---------------------- | ------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `initPayload`          | `null`             | Payload sent to Rasa when conversation starts                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| `socketUrl`            | `null`             | Socket URL                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
| `socketPath`           | `null`             | Close the chat window                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| `customData`           | `null`             | Arbitrary object sent with the socket. If using with Botfront, it must include the language (e.g. `{"language": "en"}`)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| `docViewer`            | `false`            | If you add this prop to the component or to the init script, `docViewer=true` , this will treat links in received messages as links to a document ( `.pdf .doc .xlsx` etc. ) and will open them in a popup using `https://docs.google.com/viewer` service                                                                                                                                                                                                                                                                                                                                                    |
| `title`                | `'Welcome"`        | Title shown in the header of the widget                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| `subtitle`             | `null`             | Subtitle shown under the title in the header of the widget                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
| `inputTextFieldHint`   | `"Type a message"` | User message input field placeholder                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| `hideWhenNotConnected` | `true`             | If `true` the widget will hide when the connection to the socket is lost                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
| `connectOn`            | `"mount"`          | This prop lets you choose when the widget will try connecting to the server. By default, it tries connecting as soon as it mounts. If you select `connectOn='open'` it will only attempt connection when the widget is opened. it can only take the values `mount` and `open`.                                                                                                                                                                                                                                                                                                                               |
| `onSocketEvent`        | `null`             | call custom code on a specific socket event                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
| `embedded`             | `false`            | Set to `true` if you want to embed the in a web page. The widget will always be open and the `initPayload` will be triggered immediately                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
| `showFullScreenButton` | `false`            | Show a full screen toggle                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| `displayUnreadCount`   | `false`            | Path to an image displayed on the launcher when the widget is closed                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| `showMessageDate`      | `false`            | Show message date. Can be overriden with a function: `(timestamp, message) => return 'my custom date'`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
| `customMessageDelay`   | See below          | This prop is a function, the function take a message string as an argument. The defined function will be called everytime a message is received and the returned value will be used as a milliseconds delay before displaying a new message.                                                                                                                                                                                                                                                                                                                                                                 |
| `params`               | See below          | Essentially used to customize the image size, also used to change storage options.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
| `storage`              | `"local"`          | ⚠️ This is not a prop, it has to be passed inside the params object above. <br> Specifies the storage location of the conversation state in the browser. `"session"` defines the state to be stored in the session storage. The session storage persists on reload of the page, and is cleared after the browser or tab is closed, or when `sessionStorage.clear()`is called. `"local"` defines the state to be stored in the local stoage. The local storage persists after the the browser is closed, and is cleared when the cookies of the browser are cleared, or when `localStorage.clear()`is called. |
| `customComponent`      | `null`             | Custom component to be used with custom responses. E.g.: `customComponent={ (messageData) => (<div>Custom React component</div>)}`. Please note that this can only be used if you call the webchat from a React application as you can't write a component in pure Javscript.                                                                                                                                                                                                                                                                                                                                |
| `onWidgetEvent`        | `{}`               | call custom code on a specific widget event ( `onChatOpen`, `onChatClose`, `onChatHidden`, are available for now ), add a function to the desired object property in the props to have it react to the event.                                                                                                                                                                                                                                                                                                                                                                                                |
| ---                    |

### Additional Examples

##### `customMessageDelay`

```javascript
(message) => {
    let delay = message.length * 30;
    if (delay > 2 * 1000) delay = 3 * 1000;
    if (delay < 400) delay = 1000;
    return delay;
};
```

##### `onSocketEvent`

```jsx
onSocketEvent={{
  'bot_uttered': () => console.log('the bot said something'),
  'connect': () => console.log('connection established'),
  'disconnect': () => doSomeCleanup(),
}}
```

---

## Other features

### Tooltips

Text messages received when the widget is closed will be shown as a tooltip.

### Sending a message on page load

When reconnecting to an existing chat session, the bot will send a message contained in the localStorage key specified by the `NEXT_MESSAGE` constant. The message should be stringified JSON with a `message` property describing the message and an `expiry` property set to a UNIX timestamp in milliseconds after which this message should not be sent. This is useful if you would like your bot to be able to offer your user to navigate around the site.

### Sending a payload from your React app

```jsx
function myComponent() {
    const webchatRef = useRef(null);

    // triggered when something happens in your app
    function callback() {
        if (webchatRef.current && webchatRef.current.sendMessage) {
            webchatRef.current.sendMessage('/myIntent{"entityName":"value"}');
        }
    }

    return <RasaWebchat ref={webchatRef} />;
}
```

The payload can be any message that the user would normally send, but if you want to force an intent and maybe some entities, you can use that format
`/myIntent{"entity1":"value1","entity2":"value2"}`

### Backends

The widget can be used with any backend but is primarily designed to be used with [Rasa](https://github.com/rasaHQ/rasa) or [Botfront](https://github.com/botfront/botfront).

### Rasa

Use the `socketio` channel: See [instructions in the Rasa documentation](https://rasa.com/docs/core/connectors/#socketio-connector)

If you want to process `customData` in Rasa you have to [create a new channel](https://rasa.com/docs/core/connectors/#custom-channels). Use channel `rasa_core.channels.socketio` as a template for your new channel. In this channel, `customData` can be retrieved via `data['customData']`. Then you can modify `sender_id`, save `customData` to the database, fill slots or whatever you need to with your additional data.

### Botfront

The Rasa Webchat is developped by the [Botfront](https://botfront.io) team and it works with Botfront. If your bot is multilingual, make sure to specificy the current language in the `customData` prop. E.g. `customData={{language: 'en'}}`. See in [Botfront docs](https://botfront.io/docs/channels/webchat/) for more details.

### Styles

From version 0.8 we started prefixing all css classes, if you already had css styling for the widget, just prepend `rw-` to all the classes you changed.

hierarchy:

```
.rw-conversation-container
  |-- .rw-header
        |-- .rw-title
        |-- .rw-close-function
        |-- .rw-loading
  |-- .rw-messages-container
        |-- .rw-message
              |-- .rw-client
              |-- .rw-response
        |-- .rw-replies
              |-- .rw-reply
              |-- .rw-response
        |-- .rw-snippet
              |-- .rw-snippet-title
              |-- .rw-snippet-details
              |-- .rw-link
        |-- .rw-imageFrame
        |-- .rw-videoFrame
  |-- .rw-sender
        |-- .rw-new-message
        |-- .rw-send
```

| Class                      | Description                                                        |
| -------------------------- | ------------------------------------------------------------------ |
| .rw-widget-container       | The div containing the chatbox of the default version              |

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.1](https://https///compare/v1.0.0...v1.0.1) (2021-04-15)

### Bug Fixes

-   readme typo ([f0fa78c](https://https///commit/f0fa78c048d71e52b8a8d2c863e8e2269f2fb834))
-   socket io ([4c1562d](https://https///commit/4c1562d5d62df7f70e12910edcbe093e4dc48e80))

### [1.0.1](https://https///compare/v1.0.0...v1.0.1) (2021-04-15)

### Bug Fixes

-   readme typo ([f0fa78c](https://https///commit/f0fa78c048d71e52b8a8d2c863e8e2269f2fb834))
-   socket io ([4c1562d](https://https///commit/4c1562d5d62df7f70e12910edcbe093e4dc48e80))

## [1.0.0](https://https///compare/v0.11.12...v1.0.0) (2021-02-01)

### [0.11.12](https://https///compare/v0.11.11...v0.11.12) (2021-01-19)

### Features

-   expose getSessionId ([c8cbda7](https://https///commit/c8cbda78ec9b746f2697aa8c279c84f1b260fd1f))

### [0.11.11](https://https///compare/v0.11.10...v0.11.11) (2020-10-21)

### [0.11.10](https://https///compare/v0.11.9...v0.11.10) (2020-10-16)

### Features

-   image size in markdown ([990f4bb](https://https///commit/990f4bb113da8ae1c29528c9b48a279d5efc8393))

### [0.11.9](https://https///compare/v0.11.8...v0.11.9) (2020-10-14)

### Bug Fixes

-   custom responses ([ae7072c](https://https///commit/ae7072c0dca010dc719d8ecb067726fe8e0e7f9e))
-   image size in markdown ([cee6286](https://https///commit/cee62866b3e61806d389d1728b32864da8f02554))

### [0.11.8](https://https///compare/v0.11.7...v0.11.8) (2020-09-28)

### Features

-   possibility to send a payload on tooltip close ([38cf679](https://https///commit/38cf67976ba634ad2a6097c6f684ff105ede04e6))

### [0.11.7](https://https///compare/v0.11.6...v0.11.7) (2020-09-16)

### Bug Fixes

-   colors on buttons using color props ([dbfebe6](https://https///commit/dbfebe63699f67c9baf5551ca0aa588ca4df76d6))

### [0.11.6](https://https///compare/v0.11.5...v0.11.6) (2020-08-27)

### Bug Fixes

-   code formatting on readme ([2a4f94c](https://https///commit/2a4f94cde3137aff1f73080ce296eadafed8bc25))
-   reload in edge cases ([db6d3d1](https://https///commit/db6d3d140db90e1e417b52c264b0a524489c4a8d))

### [0.11.5](https://https///compare/v0.11.4...v0.11.5) (2020-06-24)

### Features

-   choose tooltip placement ([aba68c8](https://https///commit/aba68c8b2ad883b85cdf8931d876d809f2beb79e))

### [0.11.4](https://https///compare/v0.11.3...v0.11.4) (2020-06-24)

### [0.11.3](https://https///compare/v0.11.2...v0.11.3) (2020-06-24)

### Bug Fixes

-   sequence with highlights ([2c69f08](https://https///commit/2c69f08930ab213c1548ec6b1d7d8fac59bae010))

### [0.11.2](https://https///compare/v0.11.1...v0.11.2) (2020-06-23)

### Bug Fixes

-   bullet points in tooltips ([2c9e92f](https://https///commit/2c9e92f9d20d672ad7d8c991ec7b3db4358bc3a7))

### [0.11.1](https://https///compare/v0.11.0...v0.11.1) (2020-06-23)

### Bug Fixes

-   webchat opening ([14f457f](https://https///commit/14f457f6a142a1fbde7db972d2f6de0b2063fa34))

## [0.11.0](https://https///compare/v0.10.4...v0.11.0) (2020-06-23)

### Features

-   auto clear storage on new version to prevent errors ([db2df1e](https://https///commit/db2df1eec7841e2bbcdc20e228c143361218d4dc))
-   placed tooltips ([6857448](https://https///commit/6857448bee206b53c9a10ed3975f6c7247070253))
-   tooltip sequences ([80f048f](https://https///commit/80f048f516b29f032337f07ff85cb787824f3e70))

### [0.10.4](https://https///compare/v0.10.3...v0.10.4) (2020-06-06)

### Features

-   customize colors ([b2650b5](https://https///commit/b2650b5e5f3aaaf7b5bae7240ddadf8d7bafa27a))

### Bug Fixes

-   minor style issue on typing indicator ([0b9eb22](https://https///commit/0b9eb228e0573db38c0b77a84571bd279cbbd6f0))

### [0.10.3](https://https///compare/v0.10.2...v0.10.3) (2020-06-01)

### Bug Fixes

-   auto scrolling ([9c91876](https://https///commit/9c91876aa15b42ba7dbde32ff071b947473776bf))

### [0.10.2](https://https///compare/v0.10.1...v0.10.2) (2020-05-28)

### Bug Fixes

-   page refresh when pressing enter on firefox ([174d9b9](https://https///commit/174d9b9b0cbff81f7fe832c5e46e5d0a313a254a))

### [0.10.1](https://https///compare/v0.10.0...v0.10.1) (2020-05-27)

## [0.10.0](https://https///compare/v0.9.1...v0.10.0) (2020-05-27)

### Features

-   message input multiline ([cef1002](https://https///commit/cef10021aee0e4a32a0fd1061dd8b3a884591623))
-   persistent button ([498f1d4](https://https///commit/498f1d4bb5118cb87c3aebcf6373481a123534d3))

### Bug Fixes

-   markdown lists when embedded mode is active ([46427d1](https://https///commit/46427d143bcd56d8f53f19e4cb07bcd50e73a53f))
-   styles when widget is in full screen mode ([74a95e9](https://https///commit/74a95e9ede83a94b1c0e985a528cf8bc3b8b4912))

### [0.9.1](https://https///compare/v0.9.0...v0.9.1) (2020-05-21)

### Bug Fixes

-   allow session re-use by re-reading session_id from local storage ([6ea3ffc](https://https///commit/6ea3ffcceb2abc47f1d6fe1cf5c0821cd69fa216))

## [0.9.0](https://https///compare/v0.8.9...v0.9.0) (2020-05-11)

### [0.8.9](https://https///compare/v0.8.8...v0.8.9) (2020-05-05)

### Features

-   carousels ([98ee08f](https://https///commit/98ee08f96fe4fec504375931c211a6db8abacdba))

### [0.8.8](https://https///compare/v0.8.7...v0.8.8) (2020-04-24)

### Features

-   allow disabling/enabling input ([333baea](https://https///commit/333baea0c1b87c6d7a99455a0902a6a0f927b20d))
-   change the color of the send button when input is not empty ([#214](https://https//undefined/issues/214)) ([bd75cc5](https://https///commit/bd75cc50b39aab935e6b54281c7d7caa87a9b4c5))

### [0.8.7](https://https///compare/v0.8.6...v0.8.7) (2020-04-16)

### Bug Fixes

-   scroll to highlight centered on mobile ([560e44b](https://https///commit/560e44b077b55d8a3038736e62bd88a040dcefe9))

### [0.8.6](https://https///compare/v0.8.5...v0.8.6) (2020-04-16)

### Features


Botfront/index.js
import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import RasaWebchatPro from './src/pro-src/rules-wrapper';
import { library } from '@fortawesome/fontawesome-svg-core';
import * as Icons from '@fortawesome/free-solid-svg-icons';

import './src/pro-src/index.css';

const iconList = Object.keys(Icons)
    .filter((key) => key !== 'fas' && key !== 'prefix')
    .map((icon) => Icons[icon]);

library.add(...iconList);

class RasaWebchatProWithRules extends React.Component {
    constructor(props) {
        super(props);
        const { connectOn } = props;
        let { withRules } = props;
        if (connectOn === 'open' && withRules === true) {
            throw new Error(
                "You can't use rules and connect on open, you have to use connect on mount"
            );
        }
        this.webchatRef = null;
        if (withRules === undefined) {
            withRules = true;
        }
        this.state = {
            propsRetrieved: !withRules,
            rulesApplied: !withRules,
        };
        this.setRef = this.setRef.bind(this);
        this.handleSessionConfirm = this.handleSessionConfirm.bind(this);
    }

    setRef(element) {
        const { innerRef } = this.props;
        if (!innerRef) {
            this.webchatRef = element;
        } else if (innerRef && innerRef.constructor && innerRef.call && innerRef.apply) {
            // if this is true, innerRef is a function and thus it's a callback ref
            this.webchatRef = element;
            innerRef(element);
        } else {
            innerRef.current = element;
        }
    }

    handleSessionConfirm(sessionObject) {
        const { innerRef } = this.props;
        this.setState({
            // The OR makes it work even without the augmented webchat channel
            propsRetrieved: { ...sessionObject.props },
        });
        if (
            ((innerRef && innerRef.current) || this.webchatRef.updateRules) &&
            sessionObject.props &&
            sessionObject.props.rules
        ) {
            setTimeout(() => {
                if (innerRef && innerRef.current) {
                    innerRef.current.updateRules(sessionObject.props.rules);
                } else {
                    this.webchatRef.updateRules(sessionObject.props.rules);
                }
            }, 100);
            this.setState({ rulesApplied: true });
        }
    }

    render() {
        const { onSocketEvent } = this.props;
        let { withRules } = this.props;
        if (withRules === undefined) {
            withRules = true;
        }
        const { propsRetrieved } = this.state;
        let propsToApply = {};
        if (propsRetrieved) propsToApply = propsRetrieved;
        delete propsToApply.rules;
        return (
            <div
                style={{ display: propsRetrieved ? undefined : 'none' }}
                className={
                    this.props.embedded || (propsToApply && propsToApply.embedded)
                        ? 'rw-pro-widget-embedded'
                        : ''
                }
            >
                <RasaWebchatPro
                    ref={this.setRef}
                    {...{
                        ...propsToApply,
                        ...this.props,
                    }}
                    onSocketEvent={
                        withRules
                            ? {
                                  session_confirm: this.handleSessionConfirm,
                                  ...onSocketEvent,
                              }
                            : { ...onSocketEvent }
                    }
                />
            </div>
        );
    }
}

export const rasaWebchatProTypes = {
    initPayload: PropTypes.string,
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    subtitle: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    protocol: PropTypes.string,
    socketUrl: PropTypes.string.isRequired,
    socketPath: PropTypes.string,
    protocolOptions: PropTypes.shape({}),
    customData: PropTypes.shape({}),
    handleNewUserMessage: PropTypes.func,
    profileAvatar: PropTypes.string,
    inputTextFieldHint: PropTypes.string,
    connectingText: PropTypes.string,
    showCloseButton: PropTypes.bool,
    showFullScreenButton: PropTypes.bool,
    hideWhenNotConnected: PropTypes.bool,
    connectOn: PropTypes.oneOf(['mount', 'open']),
    autoClearCache: PropTypes.bool,
    onSocketEvent: PropTypes.objectOf(PropTypes.func),
    fullScreenMode: PropTypes.bool,
    enableMessageTone: PropTypes.bool,
    messageToneMode: PropTypes.bool,
    badge: PropTypes.number,
    embedded: PropTypes.bool,
    // eslint-disable-next-line react/forbid-prop-types
    params: PropTypes.object,
    openLauncherImage: PropTypes.string,
    closeImage: PropTypes.string,
    docViewer: PropTypes.bool,
    customComponent: PropTypes.func,
    displayUnreadCount: PropTypes.bool,
    showMessageDate: PropTypes.bool,
    customMessageDelay: PropTypes.func,
    tooltipPayload: PropTypes.string,
    tooltipDelay: PropTypes.number,
    withRules: PropTypes.bool,
    rules: PropTypes.arrayOf(
        PropTypes.shape({
            payload: PropTypes.string.isRequired,
            text: PropTypes.string,
            trigger: PropTypes.shape({
                url: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
                timeOnPage: PropTypes.number,
                numberOfVisits: PropTypes.number,
                numberOfPageVisits: PropTypes.number,
                device: PropTypes.string,
                when: PropTypes.oneOf(['always', 'init']),
                queryString: PropTypes.arrayOf(
                    PropTypes.shape({
                        param: PropTypes.string,
                        value: PropTypes.string,
                        sendAsEntity: PropTypes.bool,
                    })
                ),
                eventListeners: PropTypes.arrayOf(
                    PropTypes.shape({
                        selector: PropTypes.string.isRequired,
                        event: PropTypes.string.isRequired,
                    })
                ),
            }),
        })
    ),
    triggerEventListenerUpdateRate: PropTypes.number,
};

RasaWebchatProWithRules.propTypes = {
    ...rasaWebchatProTypes,
    innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.shape({ current: PropTypes.object })]),
};

export const rasaWebchatProDefaultTypes = {
    title: 'Welcome',
    customData: {},
    inputTextFieldHint: 'Type a message...',
    connectingText: 'Waiting for server...',
    fullScreenMode: false,
    hideWhenNotConnected: true,
    autoClearCache: false,
    connectOn: 'mount',
    onSocketEvent: {},
    protocol: 'socketio',
    socketUrl: 'http://localhost',
    protocolOptions: {},
    badge: 0,
    embedded: false,
    params: {
        storage: 'session',
    },
    docViewer: true,
    showCloseButton: true,
    showFullScreenButton: false,
    enableMessageTone: false,
    messageToneMode: false,
    displayUnreadCount: true,
    showMessageDate: true,
    customMessageDelay: (message) => {
        let delay = message.length * 30;
        if (delay > 3 * 1000) delay = 3 * 1000;
        if (delay < 800) delay = 800;
        return delay;
    },
    tooltipPayload: null,
    tooltipDelay: 500,
    withRules: true,
    rules: null,
    triggerEventListenerUpdateRate: 500,
};

export default React.forwardRef((props, ref) => (
    <RasaWebchatProWithRules innerRef={ref} {...props} />
));

export const selfMount = (props, element = null) => {
    const load = () => {
        if (element === null) {
            const node = document.createElement('div');
            node.setAttribute('id', 'rasaWebchatPro');
            document.body.appendChild(node);
        }
        const mountElement = element || document.getElementById('rasaWebchatPro');
        const webchatPro = React.createElement(RasaWebchatProWithRules, props);
        ReactDOM.render(webchatPro, mountElement);
    };
    if (document.readyState === 'complete') {
        load();
    } else {
        window.addEventListener('load', () => {
            load();
        });
    }
};
